herbert_core/utilities/maths/noisify.m

```diff
@@ -17,40 +17,69 @@
 %           Add noise with Poisson distribution, where the mean value at
 %           a point is the value y.
 %
-%   If no input errors, e, just set e=[]
+%   >> [yout, eout] = noisify(y,e,[factor,]'maximum_value',maxval)
+%           Add noise with Gaussian distribution, calculating the standard
+%           deviation by an externally provided maximum y value.
+%           The max value is preceded by a keyword string 'maximum_value'.
+%           Typically this value will be the overall maximum value if the
+%           data is processed page by page and the maximum must be
+%           extracted before this processing.
+%           Omitting factor will use the default value 0.1.
+%
+%   If no input errors, e, just set e=[]. Note that eout will be created
+%   regardless of whether e is present, so the first overload output is
+%   probably incomplete.
+
+% deal with optional arguments
+is_poisson= false;
+fac = 0.1;
+if nargin>=2
```

Comment:
Is this `if` statement necessary? Do you not get the same behaviour by removing this `if` block and making the call on line 38 `addOptional`?

Reply:
We then have 2 optional parameters e and disp_or_fac with the possibility that e gets omitted and disp_or_fac does not. This is not a supported option from the header comments and I don't think I want to start choosing between optional parameters. If there are more than 2 parameters then e should be required. Ideally disp_or_fac should be a named parameter but that changes the existing interface.

Reply:
There are also suggestions in the code that Toby would prefer e=[] rather than omitting it.

Reply:
The `addOptional` function adds a positional argument - so there's no way that one could specify `disp_or_fact` and not `e`.

Reply:
This is not clear. All the inputParser docs say is that, combined with your/Nick's suggested code, is that e and disp_or_fac are both optional. I don't see how this form of argument parsing prevents leaving out e and using disp_or_fac - it will depend on the validation for e. It doesn't look like the normal method of arguments with defaults (ie optionals) where the first n optional arguments match.

Reply:
Ah, you're right. I'd misunderstood the parser. I hadn't realised you could also specify parameters added via `addOptional` as named arguments.